ldms/src/sampler/cxi_sampler/cxi_sampler.c

```diff
@@ -578,10 +574,63 @@ static int constructor(ldmsd_plug_handle_t handle)
 
 static void destructor(ldmsd_plug_handle_t handle)
 {
+    cxi_t cxi = ldmsd_plug_ctxt_get(handle);
+    if (!cxi)
+        return;
+
+    // handle ldms_set_t deletion
+    if (cxi->set)
+        base_set_delete(cxi->base);
+
+    if (cxi->schema)
+        base_schema_delete(cxi->base);
+
+    // Free interface names
+    if (cxi->iface_names) {
+        for (int i = 0; i < cxi->iface_count; i++) {
+            free(cxi->iface_names[i]);
+        }
+        free(cxi->iface_names);
+    }
+
+    // Free tel_path and rh_path
+    free(cxi->tel_path);
+    free(cxi->rh_path);
+
+    // Free telemetry files
+    if (cxi->tel_files) {
+        for (int i = 0; i < cxi->iface_count; i++) {
+            for (int j = 0; j < cxi->tel_files[i].count; j++) {
+                free(cxi->tel_files[i].names[j]);
+            }
+        }
+        free(cxi->tel_files);
+    }
+
+    // Free retry handler files
+    if (cxi->rh_files) {
+        for (int i = 0; i < cxi->iface_count; i++) {
+            for (int j = 0; j < cxi->rh_files[i].count; j++) {
+                free(cxi->rh_files[i].names[j]);
+            }
+        }
+        free(cxi->rh_files);
+    }
+
+    // Free LDMS records
+    if (cxi->tel_rec)
+        ldms_record_delete(cxi->tel_rec);
```

Comment:
I see why this looks like the right thing to do. We really need to fix the documentation for this function (and maybe even tweak the records API...).

Once records are added to a schema, they are owned by the schema. They will be freed when schema is freed. So calling ldms_record_delete() here could very well cause a segfault.

In fact, I would suggest removing "tel_rec and "rh_rec" from struct cxi_s. We can just use function-scope variables in create_metric_set(). After the record is attached to the schema, I don't think we ever need to (or maybe even _can_) do anything with it.

Reply:
@morrone, yes we should distinguish between a record definition and a record instance. It is confusing.

Reply:
@morrone, we need to have the record definition to create new record instances, however, we could look them up in the schema instead of caching them in the CXI instance.

Reply:
> @morrone, we need to have the record definition to create new record instances, however, we could look them up in the schema instead of caching them in the CXI instance.

We would only need the record definition again if we need to generate a new schema. To add record instances in a metric set, we only need the index number of the record from the schema. Those are recorded in cxi_s in:

```
	int tel_rec_mid;
	int rh_rec_mid;
```

Reply:
And, actually, I don't think a record definition _can_ be reused in a second schema. Because the record definition becomes owned by that one schema once it is attached to it.